In [4]:
# Load and examine the MOOC dataset
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv('/Users/le/Desktop/2025 Spring/S022 computing and data science/022 final project/S022_R_Project/data/MOOC_cleaned.csv')  # Replace with your file path


# Basic info about the dataset
print("Dataset shape:", df.shape)
print("\
Column names and types:")
print(df.dtypes)
print("\
First few rows:")
print(df.head())

Dataset shape: (260838, 18)
Column names and types:
course_id             object
userid_DI             object
viewed                 int64
explored               int64
certified              int64
final_cc_cname_DI     object
LoE_DI                object
YoB                    int64
gender                object
start_time_DI         object
last_event_DI         object
nforum_posts           int64
nevents_log          float64
nplay_video_log      float64
ndays_act_log        float64
grade_log            float64
nchapters_log        float64
nforum_posts_log     float64
dtype: object
First few rows:
                    course_id       userid_DI  viewed  explored  certified  \
0  HarvardX/CB22x/2013_Spring  MHxPC130539455       1         0          0   
1  HarvardX/CB22x/2013_Spring  MHxPC130088379       1         0          0   
2  HarvardX/CB22x/2013_Spring  MHxPC130024894       1         0          0   
3  HarvardX/CB22x/2013_Spring  MHxPC130340959       1         0          0   
4  Har

This code snippet loads a cleaned MOOC dataset and displays its shape, column names with types, and the first few rows.
- Load the dataset from a CSV file
- Print the shape of the dataset
- Print the data types of each column
- Display the first few rows of the dataset

In [5]:
# Get more detailed information about the dataset
print("Dataset summary statistics:")
print(df.describe())

print("\
Unique values in key categorical columns:")
print("Number of unique courses:", df['course_id'].nunique())
print("Number of unique users:", df['userid_DI'].nunique())
print("Countries:", df['final_cc_cname_DI'].value_counts().head(10))
print("\
Education levels:", df['LoE_DI'].value_counts())
print("\
Gender distribution:", df['gender'].value_counts())

Dataset summary statistics:
             viewed       explored      certified            YoB  \
count  260838.00000  260838.000000  260838.000000  260838.000000   
mean        0.49538       0.071297       0.025165    1983.876352   
std         0.49998       0.257321       0.156626       9.654767   
min         0.00000       0.000000       0.000000    1931.000000   
25%         0.00000       0.000000       0.000000    1980.000000   
50%         0.00000       0.000000       0.000000    1986.000000   
75%         1.00000       0.000000       0.000000    1990.000000   
max         1.00000       1.000000       1.000000    1997.000000   

        nforum_posts    nevents_log  nplay_video_log  ndays_act_log  \
count  260838.000000  260838.000000    260838.000000  260838.000000   
mean        0.014806       3.077671         1.834963       1.279803   
std         0.170862       1.765387         1.459118       0.748111   
min         0.000000       0.693147         0.693147       0.693147   
25% 

This code snippet summarizes the dataset's statistics and counts unique values for key categorical columns.
- Print summary statistics of the dataset
- Count and print unique courses
- Count and print unique users
- Display the top 10 countries by count
- Display counts of education levels
- Display counts of gender distribution

In [6]:
# Let's examine the engagement patterns and create some initial insights
print("Engagement funnel analysis:")
print("Viewed courses:", df['viewed'].sum())
print("Explored courses:", df['explored'].sum()) 
print("Certified completions:", df['certified'].sum())

print("\
Conversion rates:")
total_enrolled = len(df)
viewed_rate = df['viewed'].sum() / total_enrolled * 100
explored_rate = df['explored'].sum() / total_enrolled * 100
certified_rate = df['certified'].sum() / total_enrolled * 100

print(f"View rate: {viewed_rate:.1f}%")
print(f"Exploration rate: {explored_rate:.1f}%") 
print(f"Certification rate: {certified_rate:.1f}%")

# Check date ranges
df['start_time_DI'] = pd.to_datetime(df['start_time_DI'])
df['last_event_DI'] = pd.to_datetime(df['last_event_DI'])

print(f"\
Date range: {df['start_time_DI'].min()} to {df['last_event_DI'].max()}")
print("Courses offered:", df['course_id'].unique())

Engagement funnel analysis:
Viewed courses: 129214
Explored courses: 18597
Certified completions: 6564
Conversion rates:
View rate: 49.5%
Exploration rate: 7.1%
Certification rate: 2.5%
Date range: 2012-07-23 00:00:00 to 2013-11-17 00:00:00
Courses offered: ['HarvardX/CB22x/2013_Spring' 'HarvardX/CS50x/2012'
 'HarvardX/ER22x/2013_Spring' 'HarvardX/PH207x/2012_Fall'
 'HarvardX/PH278x/2013_Spring']


### Key Dataset Insights:
Engagement Funnel:

Viewed courses:
129214

Explored courses:
18597

Certified completions:
6564

Conversion rates:
View rate: 49.5%
Exploration rate: 7.1%
Certification rate: 2.5%